add a spec for loop unroll (#263)
microsoft/hlsl-specs: specs/language/statements.tex

\Ch{Statements}{Stmt}
\Sec{Label Statements}{Stmt.Label}

> \Sec{Attributes}{Stmt.Attr}

Author: visual of : [b161303](https://github.com/microsoft/hlsl-specs/pull/263/commits/b1613031a53abf2f1ac0d2052257b617aa98b6f7)
![image](https://github.com/microsoft/hlsl-specs/assets/1802579/00666515-df91-4557-a3b5-fe6c1195c61c)
![image](https://github.com/microsoft/hlsl-specs/assets/1802579/f6f0e993-624e-481d-b632-85db8fd291bf)

Author: visual of  https://github.com/microsoft/hlsl-specs/pull/263/commits/3d4f3984157a6f655986b3ebe3b11cbd118a938f
unroll spec:
![image](https://github.com/microsoft/hlsl-specs/assets/1802579/5eb4455e-ca7e-42f3-97f2-898baf416aea)

loop spec:
![image](https://github.com/microsoft/hlsl-specs/assets/1802579/9225bfd3-f631-4d8d-9ee7-e3611bf2d080)

Author: visual of  https://github.com/microsoft/hlsl-specs/pull/263/commits/aec93914bb9a77ed8759e6244929fb3e18ab95f2
![image](https://github.com/microsoft/hlsl-specs/assets/1802579/812fbfe3-0adb-4477-81dc-1b3444cedfd1)

\Sub{Unroll Attribute}{Stmt.Attr.Unroll}

> The Attribute \textbf{[unroll]} and \textbf{[unroll(n)]} attribute qualifiers can be used to specify that a loop (\textbf{for}, \textbf{while} and \textbf{do while} loops) can be unrolled. This attribute qualifier can be used to specify full unrolling or partial unrolling by a specified amount. This is a compiler hint and the compiler may ignore this directive.

Reviewer: We should probably aim to format tex source files to be 80 columns for similar reasons to the other specs.

> The Attribute \textbf{[unroll]} and \textbf{[unroll(n)]} attribute qualifiers can be used to specify that a loop (\textbf{for}, \textbf{while} and \textbf{do while} loops) can be unrolled. This attribute qualifier can be used to specify full unrolling or partial unrolling by a specified amount. This is a compiler hint and the compiler may ignore this directive.

Reviewer: Other parts of the spec (eg basic.tex) use `\texttt` for code / keywords.

Author: It might be my latex setup but I don't see a huge difference with italics. Its very subtle.

Italics:
![image](https://github.com/microsoft/hlsl-specs/assets/1802579/b332e2e6-b801-431d-8312-aa4767195810)

No Italics:
![image](https://github.com/microsoft/hlsl-specs/assets/1802579/87e56db1-5b95-4f4b-903d-5272300e3ec0)

In this example all it did was elongate

Reviewer: IIRC, `\texttt` changes the font to the monospaced "typewriter" one.  `\textit` is italics.

Author: oh ok  that makes more sense why the change is so subtle.

> n is the loop unrolling factor and must be a positive integral compile time constant expression. If n is not specified, the compiler determines the unrolling factor for the loop. This attribute is not compatible with the [loop] attribute.

Reviewer: Should we do something to typographically highlight that `n` is special?  I see other parts of the spec use `\texttt{n}`

Author: n italisized vs regular:
![image](https://github.com/microsoft/hlsl-specs/assets/1802579/97dda70e-959c-43a6-a707-1ab0c746a2d4)
Its such a suttle difference maybe we should bold or underline instead?

Author: I went ahead with italics assuming it just looks weird on my end.

Reviewer: I'm mostly calling out the differences in style I see between this new part of the spec and other parts.  @llvm-beanz may have more thoughts on how we should be tackling this in the doc.

Note:  The \textbf{[unroll]} attribute  must appear immediately before the loop to be affected.

\subsubsection{Unroll Examples}{Stmt.Attr.Unroll.Examples}
\begin{itemize}

> \item  Tell the compiler to unroll the above \textbf{for} loop by a factor of 2.

Reviewer: Wondering if the text here should be tweak to account for "This is a compiler hint and the compiler may ignore this directive.".

Author: I can change the woring to Requests or Suggests. We did talk about it being a hint earlier.

\begin{HLSL}
[unroll(2)]
int sum = 0;
for (int i=0; i<32; i++)
    sum += i;
\end{HLSL}
\item In this example, the compiler will determine how much to unroll the loop.
\begin{HLSL}
int i = 0;
[unroll]
do {
    ...
    i++;
} while(i < 32);
\end{HLSL}
\item Below is an  examples of \textbf{\emph{invalid}} usage of \textbf{[unroll(n)]} because n is not positive
\begin{HLSL}
[unroll(-1)]
while (...)
{
    ...
}
\end{HLSL}
\item This example is \textbf{\emph{invalid}} because the \textbf{[unroll]} attribute is used on a non-loop construct
\begin{HLSL}
[unroll]
if (...)
{
    ...
}
\end{HLSL}
\item The below example is \textbf{\emph{invalid}} because the loop unroll factor is not a compile-time constant expression.
\begin{HLSL}
    int x;
    [unroll(x)]
    for (int i=0; i<x; i++)
    {
        ...
    }
\end{HLSL}
\end{itemize}
\Sub{Loop Attribute}{Stmt.Attr.Loop}

> The Attribute \textbf{[loop]} tells the compiler to execute each iteration of the loop. In other words, its a hint to indicate a loop should not be unrolled. Therefore it is not compatible with the [unroll] attribute.

Reviewer: ```suggestion
The Attribute \textbf{[loop]} tells the compiler to execute each iteration of the loop. In other words, it's a hint to indicate a loop should not be unrolled. Therefore it is not compatible with the [unroll] attribute.
```

> \Ch{Statements}{Stmt}

Reviewer 2: Can we add a grammar definition up here? Something like:

```
\begin{grammar}
  \define{statement}\br
  labeled-statement\br
  \opt{attribute-specifier-sequence} expression-statement\br
  \opt{attribute-specifier-sequence} compound-statement\br
  \opt{attribute-specifier-sequence} iteration-statement\br
  \opt{attribute-specifier-sequence} selection-statement\br
  declaration-statement
\end{grammar}
```

Author: It Looked weird to me to not have a section header for the grammar. If thats ugly to you I can delete it.

Reviewer 2: Yea, I'd prefer no section for it. We haven't done sections for grammar entries in other places.

\Sec{Label Statements}{Stmt.Label}

> \Sec{Attributes}{Stmt.Attr}

Reviewer 2: Maybe add something here like:
```
\p The optional \textit{attribute-specifier-sequence} applies to the statement that immediately follows it.
```

\Sub{Unroll Attribute}{Stmt.Attr.Unroll}

> The \texttt{[unroll]} and \texttt{[unroll(n)]} attribute qualifiers can be used

Reviewer 2: nit: the term qualifiers has a specific meaning in C++, so we should avoid overloading its use. I think we can also term this as the _unroll_ attribute, the argument doesn't actually change what attribute it is, it just changes the behavior.

```suggestion
The \textit{unroll} attribute can be used
```

 to specify that a loop (\texttt{for}, \texttt{while} and \texttt{do while} 

> loops) can be unrolled. This attribute qualifier can be used to specify full 

Reviewer 2: I would probably specify that the _unroll_ attribute is only valid when applied to _iteration-statements_, then you don't need to explicitly list the loop keywords.

unrolling or partial unrolling by a specified amount. This is a compiler hint 
and the compiler may ignore this directive.

> \texttt{n} is the loop unrolling factor and must be a positive integral compile time 

Reviewer 2: The wording of this might be more clear as something like:
> The _unroll_ attribute may optionally have an a single argument that is an integer constant expression value greater than zero.

constant expression. If n is not specified, the compiler determines the 

> unrolling factor for the loop. This attribute is not compatible with the \texttt{[loop]}
>  attribute.

Reviewer 2: ```suggestion
unrolling factor for the loop. The \textit{unroll} may not be applied to the same \textit{iteration-statement} as the \textit{loop} attribute.
```

> Note:  The \texttt{[unroll]} attribute  must appear immediately before the loop
>  to be affected.

Reviewer 2: ```suggestion
```

> \subsubsection{Unroll Examples}{Stmt.Attr.Unroll.Examples}

Reviewer 2: The examples can be informative, but I'm not sure they're needed to illustrate the specification's details.

I think the C++ spec has a way of capturing examples like this in editor notes that are not included in the final specification. That might be a useful feature to add to capture cases like this.

Author: Will delete for now,  look into editor notes later.

\begin{itemize}
\item  Requests the compiler to unroll the above \texttt{for} loop by a factor 
of 2.
\begin{HLSL}
[unroll(2)]
int sum = 0;
for (int i=0; i<32; i++)
    sum += i;
\end{HLSL}
\item In this example, the compiler will determine how much to unroll the loop.
\begin{HLSL}
int i = 0;
[unroll]
do {
    ...
    i++;
} while(i < 32);
\end{HLSL}
\item Below is an  examples of \textbf{\emph{invalid}} usage of 
\texttt{[unroll(n)]} because n is not positive
\begin{HLSL}
[unroll(-1)]
while (...)
{
    ...
}
\end{HLSL}
\item This example is \textbf{\emph{invalid}} because the \texttt{[unroll]} 
attribute is used on a non-loop construct
\begin{HLSL}
[unroll]
if (...)
{
    ...
}
\end{HLSL}
\item The below example is \textbf{\emph{invalid}} because the loop unroll 
factor is not a compile-time constant expression.
\begin{HLSL}

>     int x;

Reviewer 3: inconsistent indentation

Author: examples are deleted.

    [unroll(x)]
    for (int i=0; i<x; i++)
    {
        ...
    }
\end{HLSL}
\end{itemize}
\Sub{Loop Attribute}{Stmt.Attr.Loop}
The Attribute \texttt{[loop]} tells the compiler to execute each iteration of 
the loop. In other words, its a hint to indicate a loop should not be 
unrolled. Therefore it is not compatible with the \texttt{[unroll]} attribute.

> Note:  The \texttt{[loop]} attribute  must appear immediately before the loop
>  to be affected.

Reviewer 2: ```suggestion
```
